R/PseudoBulk.R

```diff
@@ -594,18 +464,444 @@ FilterPseudobulkContrasts <- function(logic_list = NULL, design = NULL, use_requ
+    dplyr::mutate(gene_occurance = dplyr::n()) 
+  
+  #determine regulation-uniqueness pairs
+  tibble[tibble$regulation == "upregulated","uniqueness"] <- "up_nonunique"
+  tibble[tibble$regulation == "upregulated" & tibble$gene_occurance == 1, "uniqueness"] <- "up_unique"
+  tibble[tibble$regulation == "downregulated","uniqueness"] <- "down_nonunique"
+  tibble[tibble$regulation == "downregulated" & tibble$gene_occurance == 1, "uniqueness"] <- "down_unique"
+  
+  #geom_bar sums the n_DEG column, so fixing the sign is all that is necessary
+  tibble[tibble$regulation == "upregulated", "n_DEG"] <- 1 
+  tibble[tibble$regulation == "downregulated", "n_DEG"] <- -1 
+  tibble[tibble$regulation == "noDEG", "n_DEG"] <- 0
+  return(tibble)
+}
+
+#' @title PseudobulkingDEHeatmap
+#'
+#' @description This is a plotting function that formats the FilterPseudobulkingContrasts logic to plot a heatmap of log fold changes for a given gene space. It is recommended that you filter this gene space through some kind of feature selection, but full transcriptome heatmaps are possible.  
+#' @param seuratObj A pseudobulked Seurat object.
+#' @param geneSpace An optional vector of gene names that specify which genes should populate the heatmap. It is recommended that you perform some kind of feature selection upstream of this to select genes for plotting. 
+#' @param contrastField The primary grouping variable to display (on top of the heatmap). 
+#' @param negativeContrastValue The value of contrastField to be treated as "downregulated". 
+#' @param positiveContrastValue An optional variable to define a specific positive contrast value. While negativeContrastValue determines the log fold changes, this argument operates primarily as a filtering variable to eliminate groups and show a particular value of the contrast field.
+#' @param subgroupingVariable If there is a second relevant grouping variable, you can supply a second metadata column to group the data by. 
+#' @param assayName the name of the assay in the seurat object storing the count matrix. 
+#' @param showRowNames a passthrough variable for ComplexHeatmap controlling if the gene names should be shown or not in the heatmap. 
+#' @param sampleIdCol The metadata column denoting the variable containing the sample identifier (for grouping). 
+#' @return A list containing the filtered dataframe used for plotting and the heatmap plot itself. 
+#' @export
+
+PseudobulkingDEHeatmap <- function(seuratObj, geneSpace = rownames(seuratObj), contrastField = NULL, negativeContrastValue = NULL, positiveContrastValue = NULL, subgroupingVariable = NULL, showRowNames = FALSE, assayName = "RNA", sampleIdCol = 'cDNA_ID') {
+  
+  #subset the seuratObj according to the desired geneSpace
+  count_matrix <- GetAssayData(seuratObj, assay = assayName, layer = 'counts')
+  count_matrix <- count_matrix[geneSpace, ]
+  seuratObj <- CreateSeuratObject(counts = count_matrix, assay = assayName, meta.data = seuratObj@meta.data)
```

Comment:
Still not sure why the devel version breaks, but this could be it. I'll try not overwriting the seurat object.

Reply:
i didnt look at the logs, but could this have the effect of converting from a v4 assay to a v5 assay, and therefore hitting some other bug related to v5?

Reply:
yeah, I think it's a quirk of Assay5 that I'm not understanding.

the logs show `No feature overlap between existing object and new layer data`, which is why I jumped through the `JoinLayers()` hoops, but JoinLayers() doesn't seem to work at all after reading the object via: 

```
seuratObj <- suppressMessages(suppressWarnings(
Seurat::DietSeurat(
Seurat::UpdateSeuratObject(
readRDS("../tests/testdata/seuratOutput.rds")
))))
```

there's some chance that `seuratObj <- CreateSeuratObject()` interacts weirdly on devel instead of strictly overwriting the seurat object. I'm changing that in the next commit to see if it helps.